Sources/ScheduledTasks.php

```diff
@@ -1542,4 +1542,196 @@ function scheduled_remove_old_drafts()
 	return true;
 }
 
+/**
+ * Prune log_topics, log_boards & log_mark_boards_read.
+ * For users who haven't been active in a long time, purge these records.
+ * For users who haven't been active in a shorter time, mark boards as read,
+ * pruning log_topics.
+ */
+function scheduled_prune_log_topics()
+{
+	global $smcFunc, $sourcedir, $modSettings;
+
+	if (empty($modSettings['mark_read_max_users']))
+		return true;
+
+	// Convert to timestamps for comparison
+	$markReadCutoff = time() - $modSettings['mark_read_beyond'] * 86400;
+	$cleanupBeyond = time() - $modSettings['mark_read_delete_beyond'] * 86400;
+	$maxMembers = $modSettings['mark_read_max_users'];
+
+	// Try to prevent timeouts
+	@set_time_limit(300);
+	if (function_exists('apache_reset_timeout'))
+		@apache_reset_timeout();
+
+	// Start off by finding the records in log_boards, log_topics & log_mark_read
+	// for users who haven't been around the longest...
+
+	// Find the oldest $maxMembers in log_boards
+	$sql = "SELECT lb.id_member, m.last_login
+		FROM {db_prefix}members m
+		INNER JOIN
+		(
+			SELECT DISTINCT id_member
+			FROM {db_prefix}log_boards
+		) lb
+		ON lb.id_member = m.id_member
+		WHERE m.last_login <= {int:cutoff}
+		ORDER BY m.last_login
+		LIMIT {int:limit}";
+	$result = $smcFunc['db_query']('', $sql,
+		array(
+			'limit' => $maxMembers,
+			'cutoff' => $cleanupBeyond,
+		)
+	);
+	$members = array();
+	while($row = $smcFunc['db_fetch_assoc']($result))
+		$members[$row['id_member']] = $row['last_login'];
+	$smcFunc['db_free_result']($result);
+
+	// Find the oldest $maxMembers in log_mark_read
+	$sql = "SELECT lmr.id_member, m.last_login
+		FROM {db_prefix}members m
+		INNER JOIN
+		(
+			SELECT DISTINCT id_member
+			FROM {db_prefix}log_mark_read
+		) lmr
+		ON lmr.id_member = m.id_member
+		WHERE m.last_login <= {int:cutoff}
+		ORDER BY m.last_login
+		LIMIT {int:limit}";
+	$result = $smcFunc['db_query']('', $sql,
+		array(
+			'limit' => $maxMembers,
+			'cutoff' => $cleanupBeyond,
+		)
+	);
+	// Using id_member as key basically eats dupes...
```

Comment:
Would combine all three selects to one by using union,
```
select ... from asdf
union
select ... from asdf2
```
the union would also eliminate all dupes.

Reply:
I will test this to see if it works out to be any more efficient.  If so, will do.

Reply:
The UNION doesn't make anything more efficient, but it is more readable, so I am implementing that change.   

Regarding my other note:  2.0 tries (and fails) to prevent UNIONs, but 2.1 does not.